src/core/view/overlays/BaseStrokeToolView.cpp

```diff
@@ -0,0 +1,64 @@
+#include "BaseStrokeToolView.h"
+
+#include <cmath>
+
+#include <cairo.h>
+
+#include "model/Stroke.h"
+#include "util/Range.h"
+#include "view/Repaintable.h"
+#include "view/overlays/OverlayView.h"
+
+using namespace xoj::view;
+
+static Color strokeColorWithAlpha(const Stroke& s) {
+    Color c = s.getColor();
+    if (s.getToolType() == STROKE_TOOL_HIGHLIGHTER) {
+        c.alpha = s.getFill() == -1 ? 120U : static_cast<uint8_t>(s.getFill());
+    } else {
+        c.alpha = 255U;
+    }
+    return c;
+}
+
+BaseStrokeToolView::BaseStrokeToolView(const Repaintable* parent, const Stroke& stroke):
+        OverlayView(parent),
+        cairoOp(stroke.getToolType() == STROKE_TOOL_HIGHLIGHTER ? CAIRO_OPERATOR_MULTIPLY : CAIRO_OPERATOR_OVER),
+        strokeColor(strokeColorWithAlpha(stroke)),
+        lineStyle(stroke.getLineStyle()),
+        strokeWidth(stroke.getWidth()) {}
+
+BaseStrokeToolView::~BaseStrokeToolView() noexcept = default;
+
+auto BaseStrokeToolView::createMask() const -> Mask {
+    const double scale = this->parent->getAbsoluteScale();
+    Range visibleRange = this->parent->getVisiblePart();
+    visibleRange.addPadding(0.5 * this->strokeWidth);
+
+    const int xOffset = -static_cast<int>(std::floor(visibleRange.minX * scale));
+    const int yOffset = -static_cast<int>(std::floor(visibleRange.minY * scale));
+    const int width = static_cast<int>(std::ceil(visibleRange.maxX * scale)) + xOffset;
+    const int height = static_cast<int>(std::ceil(visibleRange.maxY * scale)) + yOffset;
+
+    return Mask(width, height, xOffset, yOffset, scale);
+}
+
+BaseStrokeToolView::Mask::Mask(int width, int height, int xOffset, int yOffset, double scale) {
+    cairo_surface_t* surf = cairo_image_surface_create(CAIRO_FORMAT_A8, width, height);
```

Comment:
For performance reasons, we might want to make it possible to use something like [`cairo_surface_create_similar()`](https://www.cairographics.org/manual/cairo-cairo-surface-t.html#cairo-surface-create-similar), otherwise we might be forcing the image buffer to be allocated in Xournal++'s heap instead of e.g. the display server or VRAM etc.

Reply:
TBH, I have no idea which of cairo's surfaces are stored where. The documentation does not mention this at all.
They also do not mention performance differences. Do you have a source (or benchmarks, or so) that shows a difference?
In an ideal world, I agree that `cairo_surface_create_similar` should give us the best surface for the job...

Performance aside, the downside of `cairo_surface_create_similar` is that we need a cairo_surface_t instance upon creation of the mask.
I guess this can be done, by initializing the mask during the first call to ShapeToolView::draw.

Reply:
I just pushed a commit making the masks through `cairo_surface_create_similar` and printing out the surface type in a terminal.
FYI, on my laptop: KDE X11 gives xlib surfaces while KDE Wayland gives image surfaces.

I'd be interested if you get different kind of surfaces depending on the OS, DE and x11 or wayland. To get this information in a terminal, you need to draw a filled highlighter shape (e.g. highlighter, with filling, rectangle).

Reply:
On Ubuntu 22.04 wayland I get `Creating mask of type: image`

Reply:
On Ubuntu 22.04 x11 I get `Creating mask of type: xlib`

Reply:
On MS Windows 10 I get `Creating mask of type: win32`

Reply:
On MacOS Catalina I get `Creating mask of type: quartz`
This may change when Gtk 3.24.35 is released and updated in `gtk-osx`.

Reply:
Thanks a lot @rolandlo for the data!

I run some benchmarks I found [here](https://openbenchmarking.org/test/pts/cairo-demos) (only relevant for X11, I think), and toyed around. It turns out, the benefits of using `cairo_surface_create_similar` over `cairo_surface_create_similar_image` (on xlib surfaces) are enormous. For instance
```
$./flower-demo-xlib --benchmark
flowers: 125.03 fps
```
vs
```
$./flower-demo-image --benchmark
flowers: 14.42 fps
```
I don't quite know where things get so much slower, probably the under-use of graphics acceleration.

Anyway, this seems to be saying that we need to replace most occurences of `cairo_image_surface_create` by a `cairo_surface_create_similar`, as @Technius was suggesting. It will be pretty complicated to know the kind of surfaces we need in all the places we use image surfaces though.

I'm getting out of the PR's scope :)